### PR TITLE
Added create service, delete service, and close service handle

### PR DESCRIPTION
**Pull request by dspurling**
Description:
Initial cut for the following msvcctl requests:
1. CreateService - This still needs work regarding the service log on account.  It appears that the password may need to be encrypted according to MS-SCMR.  It is working now with no credentials provided for the log on service account.
2. DeleteService
3. CloseServiceHandle

Testing:
Tested on Windows 7, 8.1, and 10.

### CODE REVIEW

**Comment by skhare-r7** [comment]
This looks very promising. Could you add some unit tests for these request / response classes?

File: src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        this.serviceName = serviceName;
+    }
+
+    @Override
+    public void marshal(PacketOutput packetOut) throws IOException {
+        packetOut.write(serviceHandle.getBytes());
+        if (serviceName!= null)
+            packetOut.writeString(serviceName, true);
+        else packetOut.writeNull();
+
```
Comment (skhare-r7) [inline src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java]: Could alignment be a concern here? Can you try testing with serviceName = "test" and "test1" ?

File: src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        packetOut.writeInt(access);
+
+        packetOut.writeInt(serviceConfigInfo.getServiceType().getValue());
+        packetOut.writeInt(serviceConfigInfo.getStartType().getValue());
+        packetOut.writeInt(serviceConfigInfo.getErrorControl().getValue());
+        if (serviceConfigInfo.getBinaryPathName() != null)
+            packetOut.writeString(serviceConfigInfo.getBinaryPathName(), true);
+        else packetOut.writeNull();
+
+        if (serviceConfigInfo.getLoadOrderGroup() != null)
```
Comment (skhare-r7) [inline src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java]: Same alignment concern

File: src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+            packetOut.writeInt(serviceConfigInfo.getDependencies().getBytes().length);
+        } else {
+            packetOut.writeNull();
+            //dependency size
+            packetOut.writeInt(0);
+        }
+
+        if (serviceConfigInfo.getServiceStartName() != null)
+            packetOut.writeStringRef(serviceConfigInfo.getServiceStartName(), true);
+        else packetOut.writeNull();
```
Comment (skhare-r7) [inline src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java]: Alignment..

File: src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+        packetOut.writeInt(access);
+
+        packetOut.writeInt(serviceConfigInfo.getServiceType().getValue());
+        packetOut.writeInt(serviceConfigInfo.getStartType().getValue());
+        packetOut.writeInt(serviceConfigInfo.getErrorControl().getValue());
+        if (serviceConfigInfo.getBinaryPathName() != null)
+            packetOut.writeString(serviceConfigInfo.getBinaryPathName(), true);
+        else packetOut.writeNull();
```
Comment (skhare-r7) [inline src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java]: Just tested this; the method does not seem to work with binaryPathName is set to null

File: src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        this.access = managerAccessLevel.getValue();
+        this.serviceName = serviceName;
+    }
+
+    @Override
+    public void marshal(PacketOutput packetOut) throws IOException {
+        packetOut.write(serviceHandle.getBytes());
+        if (serviceName!= null)
+            packetOut.writeString(serviceName, true);
+        else packetOut.writeNull();
```
Comment (skhare-r7) [inline src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java]: According to https://msdn.microsoft.com/en-us/library/cc245923.aspx ; the serviceName MUST not be NULL.

File: src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java (hunk 4 above)
Comment (skhare-r7) [reply src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java]: It seems this argument MUST not be NULL as well (even though it is not mentioned in MSDN)

**Comment by skhare-r7** [comment]
Could you also add the newer error codes from : https://msdn.microsoft.com/en-us/library/cc245923.aspx into SystemErrorCode.java?

**Comment by dspurling** [comment]
@skhare-r7 It looks like you already have all of the error codes from [https://msdn.microsoft.com/en-us/library/cc245923.aspx](https://msdn.microsoft.com/en-us/library/cc245923.aspx)

File: src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java (hunk 1 above)
Comment (dspurling) [reply src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java]: Alignment doesn't appear to be a problem here.

File: src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java (hunk 4 above)
Comment (dspurling) [reply src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java]: Fixed

File: src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWResponse.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import com.rapid7.client.dcerpc.io.PacketInput;
+import com.rapid7.client.dcerpc.messages.HandleResponse;
+
+import java.io.IOException;
+
+public class RCreateServiceWResponse extends HandleResponse{
+    private int tagId;
+
+    @Override
+    public void unmarshal(PacketInput packetIn) throws IOException {
```
Comment (skhare-r7) [inline src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWResponse.java]: Use unmarshalResponse instead. Unmarshal will handle the ReturnCode

File: src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java (hunk 3 above)
Comment (dspurling) [reply src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java]: No alignment problem

File: src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWResponse.java (hunk 6 above)
Comment (dspurling) [reply src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWResponse.java]: Ok, I see the difference.  Fixed in my local branch.

File: src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+            //dependency size
+            packetOut.writeInt(0);
+        }
+
+        if (serviceConfigInfo.getServiceStartName() != null)
+            packetOut.writeStringRef(serviceConfigInfo.getServiceStartName(), true);
+        else packetOut.writeNull();
+
+        if (serviceConfigInfo.getPassword() != null) {
+            // This is not working.  The password may need to be encrypted according to MS-SCMR
```
Comment (skhare-r7) [inline src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java]: How did you test this?

Comment (dspurling) [reply src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java]: I used a local account.  The account name must be formatted as ".\user".  The account must also have the service log on privilege.  Setting the account name without the password works.

I found the encryption method that should be used for the password here: https://msdn.microsoft.com/en-us/library/cc234408.aspx.  I started work on that.

File: src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java (hunk 5 above)
Comment (skhare-r7) [reply src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java]: ^ Last thing left to be addressed. Please push PR with updates

Comment (dspurling) [reply src/main/java/com/rapid7/client/dcerpc/msvcctl/messages/RCreateServiceWRequest.java]: This has been addressed locally.

**Pull request closed without merging**